lib/Dialect/ModArith/Transforms/ConvertToMac.cpp

```diff
@@ -0,0 +1,58 @@
+#include "lib/Dialect/ModArith/Transforms/ConvertToMac.h"
+
+#include "lib/Dialect/ModArith/IR/ModArithOps.h"
+#include "llvm/include/llvm/Support/Debug.h"            // from @llvm-project
+#include "mlir/include/mlir/IR/ImplicitLocOpBuilder.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+
+#define DEBUG_TYPE "mod-arith-mac"
+
+namespace mlir {
+namespace heir {
+namespace mod_arith {
+
+#define GEN_PASS_DEF_CONVERTTOMAC
+#include "lib/Dialect/ModArith/Transforms/Passes.h.inc"
+
+struct FindMac : public OpRewritePattern<mod_arith::AddOp> {
+  FindMac(mlir::MLIRContext *context)
+      : OpRewritePattern<mod_arith::AddOp>(context) {}
+
+  LogicalResult matchAndRewrite(mod_arith::AddOp op,
+                                PatternRewriter &rewriter) const override {
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+
+    auto *parent = op.getLhs().getDefiningOp();
```

Comment:
You can use `getDefiningOp<mod_arith::MulOp>` to cast and fetch at the same time, and a null check is all that is needed

https://github.com/llvm/llvm-project/blob/2dba66b8536506cb2f41a86a01e74819c40c53a4/mlir/include/mlir/IR/Value.h#L148-L151

Reply:
Cool, TIL